include/tvm/relax/op_attr_types.h

```diff
@@ -325,6 +325,50 @@ struct ReduceAttrs : public tvm::AttrsNode<ReduceAttrs> {
   }
 };  // struct ReduceAttrs
 
+/*! \brief Attributes used in reshape operator */
+struct ReshapeAttrs : public tvm::AttrsNode<ReshapeAttrs> {
+  Array<PrimExpr> new_shape;
```

Comment:
Because we now support ShapeExpr as first class, let us make reshape take Expr as the second argument and expects a Shape.

Note that this would require us to enable negative values in shape, which is OK.

Reply:
This would require some special checkings for -1

Reply:
Perhaps an implicit op, `relax.builtin.prepare_reshape_shape(src_shape, shape)`

Reply:
My consideration of making this attribute is that having negative values in ShapeExpr doesn’t sound perfectly right. -1 is only used in Reshape. And any other case where non-positive values appear in ShapeExpr or ShapeStructInfo should be regarded as invalid (I’m even thinking about adding a check to ShapeExpr constructor to enforce this).

Since the -1 inference is only for reshape, I think it’s reasonable to make Reshape treat the input specially, compared with other ops (like full/ones/zeros in #79), because those ops require normal input shape, where any non-positive value is not allowed.

Reply:
One of the main problem in this case is to have PrimExpr appear as an Attrs, which breaks our abstraction level. In this case, having constant reshape is fine. 

But we won't be able to update the prim expression if the IR get mutated.

As another alternative, perhaps we should just require this as ShapeExpr, and provide helper conversion to construct such shape Expr in reshape implementation.

```c++

Expr MakeReshape(x, Array<PrimExpr> shape) {
   ShapeExpr new_shape
   if (shape.as<Array<PrimExpr>>()) {
      // check -1, check x.shape and construct such a ShapeExpr when constructing the call 
   }
   return Call(reshape, {x, new_shape});
}

```